src/NPCManager.cpp

```diff
@@ -94,20 +94,49 @@ void NPCManager::sendToViewable(Entity *npc, void *buf, uint32_t type, size_t si
 static void npcBarkHandler(CNSocket* sock, CNPacketData* data) {
     sP_CL2FE_REQ_BARKER* req = (sP_CL2FE_REQ_BARKER*)data->buf;
 
-    // get bark IDs from task data
-    TaskData* td = Missions::Tasks[req->iMissionTaskID];
-    std::vector<int> barks;
-    for (int i = 0; i < 4; i++) {
-        if (td->task["m_iHBarkerTextID"][i] != 0) // non-zeroes only
-            barks.push_back(td->task["m_iHBarkerTextID"][i]);
+    int taskID = req->iMissionTaskID;
+    // ignore req->iNPC_ID as it is often fixated on a single npc in the region
+
+    if (Missions::Tasks.find(taskID) == Missions::Tasks.end()) {
+        std::cout << "mission task not found: " << taskID << std::endl;
+        return;
+    }
+
+    TaskData* td = Missions::Tasks[taskID];
+    auto& barks = td->task["m_iHBarkerTextID"];
+
+    Player* plr = PlayerManager::getPlayer(sock);
+    std::vector<std::pair<int32_t, int32_t>> npcLines;
+
+    for (Chunk* chunk : plr->viewableChunks) {
+        for (auto ent = chunk->entities.begin(); ent != chunk->entities.end(); ent++) {
+            if (ent->kind == EntityKind::PLAYER)
+                continue;
+
+            BaseNPC* npc = (BaseNPC*)ent->getEntity();
+            if (npc->type < 0 || npc->type >= NPCData.size())
```

Comment:
I don't think this is the correct way to validate the type (take a look at `ensureValidNPCType`). This is probably what you want (we should have a more general helper for this, really, but oh well):
```suggestion
            if (npc->type < 0 || npc->type > NPCData.back()["m_iNpcNumber"])
```

Reply:
`ensureValidNPCType` does do this, but in all instances where `NPCData` is used, we just directly use it with an array access, we don't loop and check for the correct `m_iNpcNumber`. We end up implicitly relying on the array index <-> `m_iNpcNumber` equivalence either way, so I think it's a preference. What would an ideal helper look like? Maybe we can introduce it here.

Reply:
Hmm, I actually think your way here is preferable then. We'll reevaluate the other function in a follow-up fix